### ISSUE TITLE
longEnvVars - Error: METEOR_SETTINGS are not valid JSON.

### ISSUE THREAD

**Issue by cunneen** [zodern/mup-aws-beanstalk#55]
I couldn't get the [todos example](https://github.com/meteor/todos) working with `longEnvVars : true` ; it looks like S3 is returning an XML file.

### Error Message

```
$ mup logs
=> Requesting Logs
=> Downloading Logs
i-070b488ded7a46049 nodejs/nodejs.log
-------------------------------------

> mup-todos@2.0.0 start /var/app/current
> bash ./start.sh

Long Env is enabled.
/etc/app/env.txt: line 1: syntax error near unexpected token `newline'
/etc/app/env.txt: line 1: `<?xml version="1.0" encoding="UTF-8"?>'
Node version
v8.11.2
Npm version
5.10.0
=> Starting health check server
=> Starting App
/var/app/current/programs/server/boot.js:475
}).run();
   ^

Error: METEOR_SETTINGS are not valid JSON.
    at packages/meteor.js:60:11
    at packages/meteor.js:81:4
    at packages/meteor.js:1463:3
    at /var/app/current/programs/server/boot.js:411:36
    at Array.forEach (<anonymous>)
    at /var/app/current/programs/server/boot.js:220:19
    at /var/app/current/programs/server/boot.js:471:5
    at Function.run (/var/app/current/programs/server/profile.js:510:12)
    at /var/app/current/programs/server/boot.js:470:11

npm ERR! Linux 4.14.62-65.117.amzn1.x86_64
npm ERR! argv "/opt/elasticbeanstalk/node-install/node-v6.14.3-linux-x64/bin/node" "/opt/elasticbeanstalk/node-install/node-v6.14.3-linux-x64/bin/npm" "start"
npm ERR! node v6.14.3
npm ERR! npm  v3.10.10
npm ERR! code ELIFECYCLE
npm ERR! mup-todos@2.0.0 start: `bash ./start.sh`
npm ERR! Exit status 1
```

### Full Reproduction

Set up an IAM user with an access key and secret key.

Add the IAM user to a new group.

Give the group the following permissions:

```
IAMFullAccess
AWSCloudTrailFullAccess
AWSElasticBeanstalkFullAccess
AmazonSSMFullAccess
AWSCertificateManagerFullAccess
CloudWatchEventsFullAccess
AmazonS3FullAccess
```

Set up a mongodb host (mongodb atlas) in the same region

    * Allow all IP addresses (tighten this later)
    * Create a new user
        * Permissions: `atlasAdmin@admin` , `dbAdmin@todos` , `readWrite@todos`

Set up our meteor app as normal.

```bash
git clone https://github.com/meteor/todos
cd todos
meteor npm install
```

I also needed to update the android build config

```
mkdir -p cordova-build-override/platforms/android
cat<<EOF>./cordova-build-override/platforms/android/build-extras.gradle
configurations.all {
    resolutionStrategy {
        force 'com.android.support:support-v4:27.1.0'
    }
}
EOF

```

Set up our environment variables for this particular config.

```bash
export MUP_APP_NAME='todos'
export MUP_DEPLOY_DOMAIN='todos.example.com'
export MUP_ACCESS_KEY_ID='AKIAQQQQQQQQQQ'
export MUP_SECRET_ACCESS_KEY='XXXXX/XXXX/XXXXXXXXXXXXXXX'
export MUP_SSL_DOMAIN='*.todos.example.com'
export MUP_SSL_OTHER_DOMAINS='todos.example.com my.todos.example.com'
export MUP_SSL_DOMAIN_TO_EMAIL='example.com'
# NOTE: I had to use the mongo 3.4 URL and remove retrywrites=true
export MUP_MONGO_URL='mongodb://<MONGO_USER>:<MONGO_PASSWORD>@cluster0-shard-00-00-aaaa.mongodb.net:27017,cluster0-shard-00-01-aaaa.mongodb.net:27017,cluster0-shard-00-02-aaaa.mongodb.net:27017/todos?ssl=true&replicaSet=Cluster0-shard-0&authSource=admin'
export MUP_DEPLOY_REGION='ap-southeast-1'

```

List certificates

```
AWS_ACCESS_KEY_ID=${MUP_ACCESS_KEY_ID} AWS_SECRET_ACCESS_KEY=${MUP_SECRET_ACCESS_KEY} aws --region ${MUP_DEPLOY_REGION} acm list-certificates
```

Create a new certificate if we choose, email the validation to the top-level domain

```
MUP_DVOPTIONS=''
for i in $MUP_SSL_OTHER_DOMAINS; 
do 
  MUP_DVOPTIONS="${MUP_DVOPTIONS} DomainName=${i},ValidationDomain=${MUP_SSL_DOMAIN_TO_EMAIL}"; 
done
MUP_DVOPTIONS="${MUP_DVOPTIONS}  DomainName=${MUP_SSL_DOMAIN},ValidationDomain=${MUP_SSL_DOMAIN_TO_EMAIL}"

AWS_ACCESS_KEY_ID=${MUP_ACCESS_KEY_ID} \
AWS_SECRET_ACCESS_KEY=${MUP_SECRET_ACCESS_KEY} \
aws --region ${MUP_DEPLOY_REGION} acm request-certificate \
    --domain-name ${MUP_SSL_DOMAIN} --validation-method EMAIL \
    --subject-alternative-names ${MUP_SSL_OTHER_DOMAINS} \
    --domain-validation-options ${MUP_DVOPTIONS}
```

Don't forget to check our email and click the verification links.

Set up our mup.js file from this template:

```
mkdir .deploy
cd .deploy
mup init
mv settings.json ..
ln -s ../settings.json settings.json

cat <<EOF>mup.template
module.exports = {
  app: {
      type: 'aws-beanstalk',
      name: '${MUP_APP_NAME}',
      path: '../',
      auth: {
          // IAM user's Access key ID
          id: '${MUP_ACCESS_KEY_ID}',
          // IAM user's Secret access key
          secret: '${MUP_SECRET_ACCESS_KEY}'
      },
      env: {
          ROOT_URL: 'https://${MUP_DEPLOY_DOMAIN}',
          MONGO_URL: '${MUP_MONGO_URL}'
      },

      minInstances: 2,
      maxInstances: 5,

      instanceType: 't2.small',

      forceSSL: true,
      region: '${MUP_DEPLOY_REGION}',
      gracefulShutdown: true,
      longEnvVars: true,

      // (optional, default is 3) Number of old application versions to keep
      oldVersions: 3,

      buildOptions: {
          server: 'https://${MUP_DEPLOY_DOMAIN}'
      },
	   docker: {
	       image: 'abernix/meteord:base'
	   }      
  },
  plugins: ['mup-aws-beanstalk']
}
EOF

```

Now, use our environment variables to fill the mup.js template

```
cat mup.template | envsubst > mup.js
```

Now we deploy.

```
mup deploy
```

**Comment by cunneen** [comment]
This is a show-stopper for me; the real application I want to use has a large settings file and it fails with : 

```
Service:AmazonCloudFormation, Message:Template format error: Parameter 'EnvironmentVariables' default value '[****]' length is greater than 4096.
```

Does anyone have a suggestion on how to fix the issue with `longEnvVars`

**Comment by cunneen** [comment]
Further info: after manually running the shell script from [src/assets/env.yml]() , I get this error:

```
<?xml version="1.0" encoding="UTF-8"?>
<Error><Code>PermanentRedirect</Code><Message>The bucket you are attempting to access must be addressed using the specified endpoint. Please send all future requests to this endpoint.</Message><Bucket>mup-appbuilder</Bucket><Endpoint>mup-appbuilder.s3.amazonaws.com</Endpoint><RequestId>22222222222</RequestId><HostId>QQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQQ=</HostId></Error>
```

**Comment by convexset** [comment]
Similar problem for me.

**Comment by cunneen** [comment]
@convexset try cloning my pull request into your `.deploy` folder (alongside your `mup.js` file) :

```
cd .deploy
git clone --single-branch -b issues/55-upstream https://github.com/JutoApp/mup-aws-beanstalk.git ./mup-aws-beanstalk
```

Then in your `mup.js` file, change:

```
plugins: ['mup-aws-beanstalk']
```
to

```
plugins: ['./mup-aws-beanstalk']
```

**Comment by gerwinbrunner** [comment]
So the issue is that eu-central-1 only supports the new v4 auth. The curl command in the script uses the old auth method. That's why it fails for us.

So the problematic code is the one that writes the settings environtment file:
https://github.com/zodern/mup-aws-beanstalk/blob/3f2acc3cf1be38092d2d46c2568d9e61899fe5a9/src/assets/env.yaml#L43-L58

So removing code and replacing it with the following
```
      aws s3 cp s3://mup-dhlapp-aws-production/env/1.txt /etc/app/env.txt
```
makes it work for eu-central-1

**Comment by cunneen** [comment]
Nice work @gerwinbrunner . So is the `aws` CLI already installed? And am I right in guessing that you're also setting `AWS_SECRET_ACCESS_KEY` and `AWS_ACCESS_KEY_ID` environment variables too?

**Comment by gerwinbrunner** [comment]
Both yes.
I created a pull request for your repo and for this one.

**Comment by zodern** [comment]
This is fixed in version 0.6.1. Thanks @cunneen and @gerwinbrunner for creating the pull requests.
